client/allocator/src/freeing_bump.rs

```diff
@@ -503,25 +518,53 @@ impl FreeingBumpHeapAllocator {
 	///
 	/// Returns the `bumper` from before the increase. Returns an `Error::AllocatorOutOfSpace` if
 	/// the operation would exhaust the heap.
-	fn bump(bumper: &mut u32, size: u32, heap_end: u32) -> Result<u32, Error> {
-		if *bumper + size > heap_end {
-			log::error!(
-				target: LOG_TARGET,
-				"running out of space with current bumper {}, mem size {}",
-				bumper,
-				heap_end
-			);
-			return Err(Error::AllocatorOutOfSpace)
+	fn bump(bumper: &mut u32, size: u32, memory: &mut impl Memory) -> Result<u32, Error> {
+		let required_size = u64::from(*bumper) + u64::from(size);
+
+		if required_size > memory.size() {
+			let required_pages =
+				pages_from_size(required_size).ok_or_else(|| Error::AllocatorOutOfSpace)?;
+
+			let pages = memory.pages();
+			let max_pages = memory.max_pages().unwrap_or(MAX_WASM_PAGES);
+			debug_assert!(pages <= max_pages);
+
+			if pages >= MAX_WASM_PAGES {
+				log::debug!(target: LOG_TARGET, "Trying to grow wasm pages above maximum.",);
+
+				return Err(Error::AllocatorOutOfSpace)
+			}
+
+			// Let us grow by at least pages * 2, but ensure we stay within the allowed maximum
+			// number of pages.
+			let min_grow = min(pages * 2, max_pages);
+			let next_pages = max(min_grow, required_pages);
+
+			if required_pages > max_pages {
```

Comment:
Doesn't really matter, but this `if` could be moved slightly earlier since it's checking `required_pages` and not `next_pages`.